test-suite/asianoptions.cpp

```diff
@@ -2024,6 +2095,142 @@ void AsianOptionTest::testAnalyticContinuousGeometricAveragePriceHeston() {
+        {Option::Put, 100, 100, 0, 0.05, 1.0 / 52, 0.5, 26, 0.2, "up", 2.2819},
+        {Option::Put, 100, 100, 0, 0.05, 1.0 / 52, 0.5, 26, 0.2, "down", 4.3370},
+
+        {Option::Call, 100, 110, 0, 0.05, 1.0 / 52, 0.5, 26, 0.2, "flat", 0.5515},
+        {Option::Call, 100, 110, 0, 0.05, 1.0 / 52, 0.5, 26, 0.2, "up", 0.1314},
+        {Option::Call, 100, 110, 0, 0.05, 1.0 / 52, 0.5, 26, 0.2, "down", 1.2429},
+        {Option::Put, 100, 110, 0, 0.05, 1.0 / 52, 0.5, 26, 0.2, "flat", 10.3046},
+        {Option::Put, 100, 110, 0, 0.05, 1.0 / 52, 0.5, 26, 0.2, "up", 9.8845},
+        {Option::Put, 100, 110, 0, 0.05, 1.0 / 52, 0.5, 26, 0.2, "down", 10.9960},
+
+        {Option::Call, 100, 120, 0, 0.05, 1.0 / 52, 0.5, 26, 0.2, "flat", 0.0479},
+        {Option::Call, 100, 120, 0, 0.05, 1.0 / 52, 0.5, 26, 0.2, "up", 0.0016},
+        {Option::Call, 100, 120, 0, 0.05, 1.0 / 52, 0.5, 26, 0.2, "down", 0.2547},
+        {Option::Put, 100, 120, 0, 0.05, 1.0 / 52, 0.5, 26, 0.2, "flat", 19.5541},
+        {Option::Put, 100, 120, 0, 0.05, 1.0 / 52, 0.5, 26, 0.2, "up", 19.5078},
+        {Option::Put, 100, 120, 0, 0.05, 1.0 / 52, 0.5, 26, 0.2, "down", 19.7609}};
+
+    DayCounter dc = Actual360();
+    Date today = Settings::instance().evaluationDate();
+
+    for (auto& l : cases) {
+        Time dt = (l.expiry - l.first) / (l.fixings - 1);
+        std::vector<Date> fixingDates(l.fixings);
+        fixingDates[0] = today + timeToDays(l.first, 360);
+
+        for (Size i = 1; i < l.fixings; i++) {
+            fixingDates[i] = today + timeToDays(i * dt + l.first, 360);
+        }
+
+        // Set up market data
+        ext::shared_ptr<SimpleQuote> spot(new SimpleQuote(l.underlying));
+        ext::shared_ptr<YieldTermStructure> qTS = flatRate(today, l.b + l.riskFreeRate, dc);
+        ext::shared_ptr<YieldTermStructure> rTS = flatRate(today, l.riskFreeRate, dc);
+        ext::shared_ptr<BlackVolTermStructure> volTS;
+        Volatility volSlope = 0.005;
+        if (l.slope == "flat") {
+            volTS = flatVol(today, l.volatility, dc);
+        } else if (l.slope == "up") {
+            std::vector<Volatility> volatilities(l.fixings);
+            for (Size i = 0; i < l.fixings; ++i) {
+                // Loop to fill a vector of vols from 7.5 % to 20 %
+                volatilities[i] = l.volatility - (l.fixings - 1) * volSlope + i * volSlope;
+            }
+            volTS =
+                ext::make_shared<BlackVarianceCurve>(today, fixingDates, volatilities, dc, true);
+        } else if (l.slope == "down") {
+            std::vector<Volatility> volatilities(l.fixings);
+            for (Size i = 0; i < l.fixings; ++i) {
+                // Loop to fill a vector of vols from 32.5 % to 20 %
+                volatilities[i] = l.volatility + (l.fixings - 1) * volSlope - i * volSlope;
+            }
+            volTS =
+                ext::make_shared<BlackVarianceCurve>(today, fixingDates, volatilities, dc, false);
+        } else {
```

Comment:
Can we slim this down?

Reply:
I agree it could have been neater here, but I left it with the "clumsy" structure just to make it very clear what is happening. I could've added a bunch of sign flippers or moved the if-statement inwards, so I'll at least consider it next time now that you reacted on it.